### PR TITLE
[SQLite] Set foreign keys constraints on by default

### PR DESCRIPTION
**Pull request by geelen**
Just a draft for now, because this is actually kinda complicated and I'm keen to start discussion.

SQLite, by default, doesn't enforce foreign key constraints, which is surprising. In their [docs](https://www.sqlite.org/pragma.html#pragma_foreign_keys), they say:

> As of SQLite [version 3.6.19](https://www.sqlite.org/releaselog/3_6_19.html), the default setting for foreign key enforcement is OFF. However, that might change in a future release of SQLite. The default setting for foreign key enforcement can be specified at compile-time using the [SQLITE_DEFAULT_FOREIGN_KEYS](https://www.sqlite.org/compile.html#default_foreign_keys) preprocessor macro. To minimize future problems, applications should set the foreign key enforcement flag as required by the application and not depend on the default setting.

I think it makes a lot more sense to set this to ON by default and relax it in the few cases where it's required, so this PR does that (using `SQLITE_DEFAULT_FOREIGN_KEYS=1` compile time flag).

However, this isn't straightfoward, since DO writes are implicitly wrapped in a transaction:

> [`PRAGMA foreign_keys;`] is a no-op within a transaction; foreign key constraint enforcement may only be enabled or disabled when there is no pending [BEGIN](https://www.sqlite.org/lang_transaction.html) or [SAVEPOINT](https://www.sqlite.org/lang_savepoint.html).

This means a user has to be careful about their use of `PRAGMA foreign_keys;`, since if they've implicitly created a DO transaction (by writing something or calling `transactionSync`), it will fail silently. I still think we should permit its use though, as it allows a user to do the following:

```js
class DO {
  constructor(state, env) {
    // Before any other DB operations
    state.storage.sql.exec(`PRAGMA foreign_keys=off;`)
  }
  
  // ...
}
```

This pattern would let a user fully opt-out of foreign key constraints.

The next question is how to allow a user to _temporarily_ opt-out of foreign key constraints? This PR proposes an extension to `transactionSync` that executes `PRAGMA foreign_key_check;` just before it's about to commit and rejects the "transaction" (actually a savepoint under the hood) explicitly. This avoids the nasty situation where using `PRAGMA defer_foreign_keys` causes the entire isolate to explode because the implicit transaction fails to commit.

The current version is aware if `PRAGMA foreign_keys` is already set to `off` and won't explode in that case (otherwise a user might never be able to run a transaction until they repair their data), but it still runs a query on every transaction, which might have a perf impact. Perhaps we could detect use of `defer_foreign_keys` in the authorizer and set a flag on the current transaction to speed things up?

TL;DR, this changes SQLite's default behaviour but guarantees a few things:
* By default, foreign keys are enforced and people are less likely to accidentally insert bad data.
* Users have an easy example to follow to opt-out for an entire Object by using the constructor, before any implicit transactions are in play.
* Users can use `transactionSync` and `defer_foreign_keys` to temporarily disable foreign key checks (e.g. when uploading from a SQL dump), but ensure that the data is consistent at the end of the transaction.

Thoughts?

### CODE REVIEW

File: src/workerd/api/sql-test.js
Comment (kentonv) [inline src/workerd/api/sql-test.js:1]: Can you please separate the auto-formatting of this file into a separate commit so that I can see what actually changed?

File: src/workerd/api/actor-state.c++ [hunk 1]
```diff
@@ -640,6 +640,12 @@ jsg::Value DurableObjectStorage::transactionSync(jsg::Lock& js, jsg::Function<js
     sqlite->run(SqliteDatabase::TRUSTED, kj::str("SAVEPOINT _cf_sync_savepoint_", depth));
     return js.tryCatch([&]() {
       auto result = callback(js);
+      if (!sqlite->run("SELECT * FROM pragma_foreign_keys(), pragma_foreign_key_check() WHERE foreign_keys = TRUE;").isDone()) {
```
Comment (kentonv) [inline src/workerd/api/actor-state.c++:643]: If we are really going to run this query on every transaction we should make it a prepared statement.

But can we avoid it entirely, at least in the common case where it won't apply? What if we prohibit deferred foreign key checks (in the authorizer)? Or can we use the authorizer to detect whether someone has enabled deferred checks, and only then start doing this check here?

Comment (geelen) [reply src/workerd/api/actor-state.c++:643]: Yeah that's the primary reason I pushed this as a draft PR. I want this:

> can we use the authorizer to detect whether someone has enabled deferred checks, and only then start doing this check here?

But I don't know how to pass context between those callsites? If we could do _that_, then we could also provide better error messages (instead of `not authorized`, we could give `not authorized: SQLITE_CREATE_TRIGGER disallowed for name '_cf_KV'`). Something I've wanted to add but didn't know how...

File: .prettierrc
[the remark this answers is not in the document]
Comment (geelen) [reply .prettierrc:1]: Yeah I'll back this out. Prettier effectively controls whatever auto-formatting logic each IDE has so that you don't end up having two people thrash on adding/removing semicolons/single quotes on every save, say.

But I didn't realise just how inconsistent even the single file of sql-test.js is (I tried to set prettier to make the diff as light as possible), so I'll do all that as a separate PR

File: src/workerd/api/actor-state.c++ (hunk 1 above)
Comment (geelen) [reply src/workerd/api/actor-state.c++:643]: So my other question is, where should the prepared statements be stored? Right now `ActorSqlite` has a couple (`beginTxn`,`commitTxn`), and I see `SqlStorage` having `pragmaPageCount` and `pragmaGetMaxPageCount`, and using the (currently private) `execMemoized` to store them on first use, but this is the generic `ActorState` class.

Gut feel says I want to replace `KJ_IF_MAYBE(sqlite, cache->getSqliteDatabase())` on line 633 with some kind of dynamic cast of `ActorCacheInterface` down to `ActorSqlite` so I can move `transactionSync` on to it, and store prepared statements there. Or, add `transactionSync` to `ActorCacheInterface`, and throw an error if you try to do it on a non-SQL-backed Cache... I might try that.

Comment (geelen) [reply src/workerd/api/actor-state.c++:643]: Ok so I tried and failed. Something to do with JSG and template specialisation which I don't understand (error in commit description here: https://github.com/cloudflare/workerd/commit/46936690ecaf32a06c43948ec861add8ed9cd601)

**Comment by kentonv** [comment]
Can we just prohibit deferring foreign key checks for now? I think we can do something fancier here to make this work but I'd rather not spend the time on it until we know it's worth it.

**Comment by geelen** [comment]
But without it we're back to needing the user to toggle foreign keys on/off outside of a transaction. For D1, or anything that passes through SQL and wraps things defensively in transactions, that means we're either having to inspect SQL to detect whether `PRAGMA foreign_keys=off` appears (and I'm not even sure what we'd do if we detect it), or disallow it and give people a separate API for toggling it DB-wide.

The nice thing about this proposal (and `defer_foreign_keys` in general) is that it works _with_ the idea that all statements run inside transactions, rather than against it.

**Comment by kentonv** [comment]
I don't follow. It seems like `foreign_keys` and `defer_foreign_keys` are orthogonal settings; I don't see how one replaces the other? Assuming we enable `foreign_keys=on` by default, then if someone wants to completely disable foreign key checks, then `defer_foreign_keys` doesn't really help them, does it?

**Comment by elithrar** [comment]
The dominant case where needing to relax foreign key constraints comes up during a migration or initial schema creation, especially if you want to minimize `ALTER TABLE` statements to “patch in” foreign key dependencies.

For simplicity I am all for getting the default (enforce fkeys) merged and addressing defer in a follow up so we aren’t blocked on the whole change.

**Comment by geelen** [comment]
Yeah following on from what @elithrar said, relaxing foreign key constraints is _usually_ only needed temporarily.

> Assuming we enable foreign_keys=on by default, then if someone wants to completely disable foreign key checks, then defer_foreign_keys doesn't really help them, does it?

That's why i've added the code example of setting `foreign_keys=off` in the DO's constructor. But we'd be encouraging users to use `defer_foreign_keys` as part of a batch SQL command to avoid having to think of this as a global, permanent setting.

> For simplicity I am all for getting the default (enforce fkeys) merged and addressing defer in a follow up so we aren’t blocked on the whole change.

Happy to simplify this PR and repush. The user-facing impact would be, if they use `defer_foreign_keys`, they get the right behaviour but instead of a useful/debuggable error message, they'll get a generic "Object storage was reset" error. But I can fix that in a follow-up PR.

**Comment by geelen** [comment]
Alright I've updated the PR, removed the fanciness with `foreign_key_check`, and updated the tests to show the current behaviour with  `defer_foreign_keys` exploding the isolate.

I've moved the smarter error message logic to a new branch: https://github.com/cloudflare/workerd/commit/e429aabd92929c42bd52bd5e81ec602bbe7064a9

**Comment by elithrar** [comment]
Thanks @geelen. If we have to, we probably end up with:

1. (This PR) enable fkeys and blow up
2. (Possible intermediate) don’t blow up; return useful error
3. Support defer foreign key 

cc @byule

**Comment by geelen** [comment]
Actually it's just step 1 and 2, we already support `defer_foreign_keys`.

The problem with isolates exploding is if you use `defer_foreign_keys` to insert some bad data but then _fail to make the data correct by the time your transaction ends_. If they leave it in a valid state, it's all fine, see: https://github.com/cloudflare/workerd/pull/794/files#diff-922c2bfbc00da6e5b434867a3b150eacbc34a56bb123232128a0cf4dcfec7e34R598-R604

If you don't use `defer_foreign_keys`, invalid data fails at the point of insert like any invalid statement (as of this PR). So you get a good error message. The problem is that D1 won't support any way to disable foreign keys globally, so we'll need to instruct people to use defer when restoring a SQL dump.

**Review by kentonv — APPROVED** [review]

File: src/workerd/api/sql-test.js [hunk 2]
```diff
@@ -1,13 +1,28 @@
 import * as assert from 'node:assert'
 
-function requireException(callback, expectStr) {
+function requireException(callback, expectStr, errCheck = () => {}) {
```
Comment (jasnell) [inline src/workerd/api/sql-test.js:3]: Nit: since you're importing `node:assert`... you could be using `assert.throws(...)` instead, perhaps?

File: src/workerd/api/sql-test.js [hunk 3]
```diff
@@ -1,13 +1,28 @@
...
+  }
+  throw new Error(`Expected exception '${expectStr}' but none was thrown`);
+}
+
+async function requireAsyncException(callback, expectStr, errCheck = () => {}) {
+  try {
+    await callback();
+  } catch (err) {
+    const errStr = `${err}`;
+    if (!errStr.includes(expectStr)) {
```
Comment (jasnell) [inline src/workerd/api/sql-test.js:22]: Could use `assert.strictEqual(...)`

File: src/workerd/api/sql-test.js (hunk 2 above)
Comment (geelen) [reply src/workerd/api/sql-test.js:3]: Will fix this sort of thing up in a separate PR
